Preserve types for serialization (#1932)
mono/linker: docs/serialization.md

# Serialization

> The linker cannot analyze the patterns typically used by reflection-based serializers. Such serializers should be annotated with `RequiresUnreferencedCodeAttribute`, and using them in a trimmed app will likely not work (or will work unpredictably). Set `SuppressTrimAnalysisWarnings` to `false` to see static analysis warnings for these patterns.

Reviewer: Is the `SuppressTrimAnalysisWarnings` part necessary now?

Author: I initially wanted to mention it in case people run into issues for xamarin-android apps, where warnings are suppressed by default, but I changed it to just point to the docs for warnings which should stay up-to-date wrt defaults.

As a last resort, the linker does have limited heuristics that can be enabled to keep _some_ of the types and members required for serialization, but this provides no correctness guarantees; apps which use reflection-based serializers are still considered "broken" as far as the static analysis can tell, and it is up to you to make sure that the app works as intended.

## History

The linker has historically been used for Xamarin scenarios that use reflection-based serializers like XmlSerializer, since before the introduction of the trim analysis warnings. There were limited heuristics to satisfy some simple uses of serializers. To provide backwards compatibility for such scenarios, the linker has built-in heuristics that makes some simple cases "just work", albeit in an opaque and unpredictable way.

Consider disabling this behavior by passing `--disable-serialization-discovery` if possible, but it may be necessary when using legacy serializers that don't provide source generators or a similar solution that is statically analyzable. The following is a description of the heuristics for anyone who is unfortunate enough to have to rely on this behavior.

> ## Heuristics

Reviewer 2: I think we should also mention that the logic only kicks in if the .ctor for a given serializer is marked. It's kind of obvious for apps which use serialization, but I think it's worth mentioning for linker documentation purposes.

Author: Definitely - I forgot to update it with the recent changes.

Author: Updated, PTAL

Serialization discovery is enabled by default, and can be disabled by passing `--disable-serialization-discovery`. There are three parts to the heuristics:
- Root type discovery: logic to discover types which are considered for serialization
- Type graph: recursive logic to build a set of types to consider for serialization, starting from the roots
- Preservation logic: what the linker does with the discovered types

## Root type discovery

The heuristics will discover types that satisfy _all_ of the following criteria:

- The type is used

>   There must be a statically discoverable reference to the type. In other words, if running the linker without the serialization heuristics removes a given type, then the heuristics will not keep it either.

Reviewer 2: ```suggestion
  There must be a statically discoverable reference to the type. In other words, if running the linker without the serialization heuristics removes a given type, then the heuristics will not even consider such type to become a root.
```

We might keep more types due to the type graph phase.

- The type or one of its members must be attributed with a serializer-specific attribute.

  See the sections below about the attributes you can use for each serializer. The attribute must be present on the type, or one of its defined members, including fields/properties/methods/events, public or private, though the serializers may not define attributes that can be placed on all member kinds. It is not enough to place an attribute on a base type or on a member defined by a base type.

Note that passing a type directly to a serializer constructor is _not_ enough to keep it. We do not use dataflow to discover types. For example:

```csharp

> new XmlSerializer (typeof (RootType)); // Will not keep RootType

Reviewer 2: ```suggestion
new XmlSerializer (typeof (RootType)); // Will not consider RootType as one of the root types
```

Linker will preserve the RootType since it's referenced.

```

> This pattern will not necessarily keep the type passed into the constructor, even though it is statically analyzable in theory.

Reviewer 2: ```suggestion
This pattern will not consider the type passed into the constructor as a root type, even though it is statically analyzable in theory.
```

  - _except_ `XmlIgnoreAttribute`

### DataContractSerializer attributes

On types:
- `System.Runtime.Serialization.DataContractAttribute`

On properties, fields, or events:
- `System.Runtime.Serialization.DataMemberAttribute`

## Type graph

Starting from the discovered root types, the heuristics will recursively discover the following types:

- Base types
  - including generic argument types
- Types of public instance properties defined on the type
  - a property is considered public if it has a public getter or setter
  - including public properties of the base type
  - including generic argument types
- Types of public instance fields defined on the type
  - including public fields of the base type
  - including generic argument types

Note that the types of implemented interfaces are not necessarily discovered.

## Preservation logic

For each discovered type (including root types and the recursive type graph), the linker marks the type and the following members:

- Public instance properties
  - including public or private getters and setters for such properties
- Public instance fields
- Public parameterless instance constructors

> Note that in general, private members and static members are not preserved, nor are methods or events (other than the mentioned constructor).

Reviewer 2: What happens if I put `[DataMember]` on a private field - I would assume we preserve it... but this sentence seems to say otherwise.

Author: Currently it causes the type to be considered a root, but the field won't actually be kept. I think I can fix this without building serializer-specific logic into the recursive phase by marking such members up-front.

Author: Fixed in the latest iteration (see the commit comments and doc updates)
